File: app/src/test/java/org/schabi/newpipe/settings/tabs/TabsJsonHelperTest.java
package org.schabi.newpipe.settings.tabs;

import com.grack.nanojson.JsonArray;
import com.grack.nanojson.JsonObject;
import com.grack.nanojson.JsonParser;
import com.grack.nanojson.JsonParserException;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static java.util.Objects.requireNonNull;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class TabsJsonHelperTest {
    private static final String JSON_TABS_ARRAY_KEY = "tabs";
    private static final String JSON_TAB_ID_KEY = "tab_id";

    @Test
    public void testEmptyAndNullRead() throws TabsJsonHelper.InvalidJsonException {
        final List<Tab> defaultTabs = TabsJsonHelper.getDefaultTabs();

        final String emptyTabsJson = "{\"" + JSON_TABS_ARRAY_KEY + "\":[]}";
        List<Tab> items = TabsJsonHelper.getTabsFromJson(emptyTabsJson);
        assertEquals(items, defaultTabs);

        final String nullSource = null;
        items = TabsJsonHelper.getTabsFromJson(nullSource);
        assertEquals(items, defaultTabs);
    }

    @Test
    public void testInvalidIdRead() throws TabsJsonHelper.InvalidJsonException {
        final int blankTabId = Tab.Type.BLANK.getTabId();
        final String emptyTabsJson = "{\"" + JSON_TABS_ARRAY_KEY + "\":[" +
                "{\"" + JSON_TAB_ID_KEY + "\":" + blankTabId + "}," +
                "{\"" + JSON_TAB_ID_KEY + "\":" + 12345678 + "}" +
                "]}";
        final List<Tab> items = TabsJsonHelper.getTabsFromJson(emptyTabsJson);

        assertEquals("Should ignore the tab with invalid id", 1, items.size());
        assertEquals(blankTabId, items.get(0).getTabId());
    }

    @Test
    public void testInvalidRead() {
        final List<String> invalidList = Arrays.asList(
                "{\"notTabsArray\":[]}",
                "{invalidJSON]}",
                "{}"
        );

        for (String invalidContent : invalidList) {
            try {
                TabsJsonHelper.getTabsFromJson(invalidContent);

                fail("didn't throw exception");
            } catch (Exception e) {
                boolean isExpectedException = e instanceof TabsJsonHelper.InvalidJsonException;
                assertTrue("\"" + e.getClass().getSimpleName() + "\" is not the expected exception", isExpectedException);
            }
        }
    }

    @Test
    public void testEmptyAndNullSave() throws JsonParserException {
        final List<Tab> emptyList = Collections.emptyList();
        String returnedJson = TabsJsonHelper.getJsonToSave(emptyList);
        assertTrue(isTabsArrayEmpty(returnedJson));

        final List<Tab> nullList = null;
        returnedJson = TabsJsonHelper.getJsonToSave(nullList);
        assertTrue(isTabsArrayEmpty(returnedJson));
    }

    private boolean isTabsArrayEmpty(String returnedJson) throws JsonParserException {
        JsonObject jsonObject = JsonParser.object().from(returnedJson);
        assertTrue(jsonObject.containsKey(JSON_TABS_ARRAY_KEY));
        return jsonObject.getArray(JSON_TABS_ARRAY_KEY).size() == 0;
    }

    @Test
    public void testSaveAndReading() throws JsonParserException {
        // Saving
        final Tab.BlankTab blankTab = new Tab.BlankTab();
        final Tab.DefaultKioskTab defaultKioskTab = new Tab.DefaultKioskTab();
        final Tab.SubscriptionsTab subscriptionsTab = new Tab.SubscriptionsTab();
        final Tab.ChannelTab channelTab = new Tab.ChannelTab(666, "https://example.org", "testName");
        final Tab.KioskTab kioskTab = new Tab.KioskTab(123, "trending_key");

        final List<Tab> tabs = Arrays.asList(blankTab, defaultKioskTab, subscriptionsTab, channelTab, kioskTab);
        final String returnedJson = TabsJsonHelper.getJsonToSave(tabs);

        // Reading
        final JsonObject jsonObject = JsonParser.object().from(returnedJson);
        assertTrue(jsonObject.containsKey(JSON_TABS_ARRAY_KEY));
        final JsonArray tabsFromArray = jsonObject.getArray(JSON_TABS_ARRAY_KEY);

        assertEquals(tabs.size(), tabsFromArray.size());

        final Tab.BlankTab blankTabFromReturnedJson = requireNonNull((Tab.BlankTab) Tab.from(((JsonObject) tabsFromArray.get(0))));
        assertEquals(blankTab.getTabId(), blankTabFromReturnedJson.getTabId());

        final Tab.DefaultKioskTab defaultKioskTabFromReturnedJson = requireNonNull((Tab.DefaultKioskTab) Tab.from(((JsonObject) tabsFromArray.get(1))));
        assertEquals(defaultKioskTab.getTabId(), defaultKioskTabFromReturnedJson.getTabId());

        final Tab.SubscriptionsTab subscriptionsTabFromReturnedJson = requireNonNull((Tab.SubscriptionsTab) Tab.from(((JsonObject) tabsFromArray.get(2))));
        assertEquals(subscriptionsTab.getTabId(), subscriptionsTabFromReturnedJson.getTabId());

        final Tab.ChannelTab channelTabFromReturnedJson = requireNonNull((Tab.ChannelTab) Tab.from(((JsonObject) tabsFromArray.get(3))));
        assertEquals(channelTab.getTabId(), channelTabFromReturnedJson.getTabId());
        assertEquals(channelTab.getChannelServiceId(), channelTabFromReturnedJson.getChannelServiceId());
        assertEquals(channelTab.getChannelUrl(), channelTabFromReturnedJson.getChannelUrl());
        assertEquals(channelTab.getChannelName(), channelTabFromReturnedJson.getChannelName());

        final Tab.KioskTab kioskTabFromReturnedJson = requireNonNull((Tab.KioskTab) Tab.from(((JsonObject) tabsFromArray.get(4))));
        assertEquals(kioskTab.getTabId(), kioskTabFromReturnedJson.getTabId());
        assertEquals(kioskTab.getKioskServiceId(), kioskTabFromReturnedJson.getKioskServiceId());
        assertEquals(kioskTab.getKioskId(), kioskTabFromReturnedJson.getKioskId());
    }
}

File: app/src/test/java/org/schabi/newpipe/settings/tabs/TabsJsonHelperTest.kt
package org.schabi.newpipe.settings.tabs

import kotlinx.serialization.json.*
import java.util.Objects.requireNonNull
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

class TabsJsonHelperTest {

    @Test
    @Throws(TabsJsonHelper.InvalidJsonException::class)
    fun testEmptyAndNullRead() {
        val defaultTabs = TabsJsonHelper.getDefaultTabs()

        val emptyTabsJson = "{\"$JSON_TABS_ARRAY_KEY\":[]}"
        var items = TabsJsonHelper.getTabsFromJson(emptyTabsJson)
        assertEquals(items, defaultTabs)

        val nullSource: String? = null
        items = TabsJsonHelper.getTabsFromJson(nullSource)
        assertEquals(items, defaultTabs)
    }

    @Test
    @Throws(TabsJsonHelper.InvalidJsonException::class)
    fun testInvalidIdRead() {
        val blankTabId = Tab.Type.BLANK.tabId
        val emptyTabsJson = "{\"$JSON_TABS_ARRAY_KEY\":[" +
            "{\"$JSON_TAB_ID_KEY\":$blankTabId}," +
            "{\"$JSON_TAB_ID_KEY\":12345678}" + "]}"
        val items = TabsJsonHelper.getTabsFromJson(emptyTabsJson)

        assertEquals("Should ignore the tab with invalid id", 1, items.size)
        assertEquals(blankTabId, items[0].tabId)
    }

    @Test
    fun testInvalidRead() {
        val invalidList = listOf(
            "{\"notTabsArray\":[]}",
            "{invalidJSON]}",
            "{}"
        )

        for (invalidContent in invalidList) {
            try {
                TabsJsonHelper.getTabsFromJson(invalidContent)

                fail("didn't throw exception")
            } catch (e: Exception) {
                val isExpectedException = e is TabsJsonHelper.InvalidJsonException
                assertTrue(
                    "\"" + e.javaClass.simpleName + "\" is not the expected exception",
                    isExpectedException
                )
            }
        }
    }

    @Test
    @Throws(Exception::class)
    fun testEmptyAndNullSave() {
        val emptyList = emptyList<Tab>()
        var returnedJson = TabsJsonHelper.getJsonToSave(emptyList)
        assertTrue(isTabsArrayEmpty(returnedJson))

        val nullList: List<Tab>? = null
        returnedJson = TabsJsonHelper.getJsonToSave(nullList)
        assertTrue(isTabsArrayEmpty(returnedJson))
    }

    @Throws(Exception::class)
    private fun isTabsArrayEmpty(returnedJson: String): Boolean {
        val jsonObject = Json.parseToJsonElement(returnedJson).jsonObject
        assertTrue(jsonObject.containsKey(JSON_TABS_ARRAY_KEY))
        return jsonObject.getArray(JSON_TABS_ARRAY_KEY).isEmpty()
    }

    @Test
    @Throws(Exception::class)
    fun testSaveAndReading() {
        // Saving
        val blankTab = Tab.BlankTab()
        val defaultKioskTab = Tab.DefaultKioskTab()
        val subscriptionsTab = Tab.SubscriptionsTab()
        val channelTab = Tab.ChannelTab(666, "https://example.org", "testName")
        val kioskTab = Tab.KioskTab(123, "trending_key")
        val feedGroupTab = Tab.FeedGroupTab(1L, "x", 123)

        val tabs = listOf(
            blankTab,
            defaultKioskTab,
            subscriptionsTab,
            channelTab,
            kioskTab,
            feedGroupTab
        )
        val returnedJson = TabsJsonHelper.getJsonToSave(tabs)

        // Reading
        val jsonObject = Json.parseToJsonElement(returnedJson).jsonObject
        assertTrue(jsonObject.containsKey(JSON_TABS_ARRAY_KEY))
        val jsonTabs = jsonObject["tabs"]?.jsonArray
        assertNotNull(jsonTabs)
        assertEquals(tabs.size.toLong(), jsonTabs!!.size.toLong())

        val blankTabFromReturnedJson = requireNonNull(
            Tab.from(jsonTabs[0] as JsonObject) as Tab.BlankTab
        )
        assertEquals(blankTab.tabId, blankTabFromReturnedJson.tabId)

        val defaultKioskTabFromReturnedJson = requireNonNull(
            Tab.from(tabsFromArray[1] as JsonObject) as Tab.DefaultKioskTab
        )
        assertEquals(defaultKioskTab.tabId, defaultKioskTabFromReturnedJson.tabId)

        val subscriptionsTabFromReturnedJson = requireNonNull(
            Tab.from(tabsFromArray[2] as JsonObject) as Tab.SubscriptionsTab
        )
        assertEquals(subscriptionsTab.tabId, subscriptionsTabFromReturnedJson.tabId)

        val channelTabFromReturnedJson = requireNonNull(
            Tab.from(tabsFromArray[3] as JsonObject) as Tab.ChannelTab
        )
        assertEquals(channelTab.tabId, channelTabFromReturnedJson.tabId)
        assertEquals(channelTab.channelServiceId, channelTabFromReturnedJson.channelServiceId)
        assertEquals(channelTab.channelUrl, channelTabFromReturnedJson.channelUrl)
        assertEquals(channelTab.channelName, channelTabFromReturnedJson.channelName)

        val kioskTabFromReturnedJson = requireNonNull(
            Tab.from(tabsFromArray[4] as JsonObject) as Tab.KioskTab
        )
        assertEquals(kioskTab.tabId, kioskTabFromReturnedJson.tabId)
        assertEquals(kioskTab.kioskServiceId, kioskTabFromReturnedJson.kioskServiceId)
        assertEquals(kioskTab.kioskId, kioskTabFromReturnedJson.kioskId)

        val grpTabFromReturnedJson = requireNonNull(
            Tab.from(tabsFromArray[5] as JsonObject) as Tab.FeedGroupTab
        )
        assertEquals(feedGroupTab.tabId, grpTabFromReturnedJson.tabId)
        assertEquals(feedGroupTab.feedGroupId, grpTabFromReturnedJson.feedGroupId)
        assertEquals(feedGroupTab.iconId, grpTabFromReturnedJson.iconId)
        assertEquals(feedGroupTab.feedGroupName, grpTabFromReturnedJson.feedGroupName)
    }

    companion object {
        private const val JSON_TABS_ARRAY_KEY = "tabs"
        private const val JSON_TAB_ID_KEY = "tab_id"
    }
}
